### PR TITLE
Read and write union types

### PR DESCRIPTION
**Pull request by henridf**
This PR defines the union type and adds support for reading and writing unions in zng and bzng.

### CODE REVIEW

File: zng/resolver/context.go [hunk 1]
```diff
@@ -383,16 +415,14 @@ func (c *Context) parseColumn(in string) (string, zng.Column, error) {
...
-func (c *Context) parseSetTypeBody(in string) (string, zng.Type, error) {
+// parseTypeList parses a type list of the form "[type1,type2,type3]".
+func (c *Context) parseTypeList(in string) (string, []zng.Type, error) {
+	var types []zng.Type
 	rest, ok := match(in, "[")
 	if !ok {
 		return "", nil, zng.ErrTypeSyntax
 	}
 	in = rest
-	var types []zng.Type
```
Comment (nwt) [inline zng/resolver/context.go]: Nit: Maybe keep this here so it stays closer to the first use of `types`.

File: zng/type.go [hunk 2]
```diff
@@ -108,6 +109,7 @@ const (
 	TypeDefRecord = 0x80
 	TypeDefArray  = 0x81
 	TypeDefSet    = 0x82
+	TypeDefUnion  = 0x84
 	TypeDefAlias  = 0x83
```
Comment (nwt) [inline zng/type.go]: Nit: Maybe keep these sorted by value, as they currently are.

File: zng/union.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+func (t *TypeUnion) ID() int {
+	return t.id
+}
+
+func (t *TypeUnion) SetID(id int) {
+	t.id = id
+}
+
+func (t *TypeUnion) TypeIndex(index int) (Type, error) {
+	if index < 0 || index+1 > len(t.Types) {
```
Comment (nwt) [inline zng/union.go]: Nit: I think `index >= len(t.Types)` would be clearer.

File: zng/union.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	index, err := strconv.Atoi(string(in[0:c]))
+	if err != nil {
+		return nil, err
+	}
+	typ, err := t.TypeIndex(index)
+	if err != nil {
+		return nil, err
+	}
+	out := zcode.AppendUvarint(zcode.Bytes{}, uint64(index))
```
Comment (nwt) [inline zng/union.go]: Nit: s/zcode.Bytes{}/nil/ unless I'm missing something and that won't compile.

File: zng/union.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	val, err := typ.Parse(in[c+1:])
+	if err != nil {
+		return nil, err
+	}
+	out = append(out, val...)
+	return out, nil
+}
+
+func (t *TypeUnion) StringOf(zv zcode.Bytes, ofmt OutFmt, inContainer bool) string {
+	index, n := binary.Uvarint(zv)
```
Comment (nwt) [inline zng/union.go]: Sadly, n can be less than or equal to zero, signaling an error. We ought to handle it.

File: zng/union.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		return "ERR"
+	}
+	zv = zv[n:]
+	s := strconv.FormatUint(index, 10) + ":"
+	return s + innerType.StringOf(zv, ofmt, false)
+}
+
+func (t *TypeUnion) Marshal(zv zcode.Bytes) (interface{}, error) {
+	index, n := binary.Uvarint(zv)
```
Comment (nwt) [inline zng/union.go]: Handle n <= 0 here, too.

File: zng/resolver/context.go (hunk 1 above)
Comment (henridf) [reply zng/resolver/context.go]: done

File: zng/type.go (hunk 2 above)
Comment (henridf) [reply zng/type.go]: oops yes!

Comment (henridf) [reply zng/type.go]: actually double-oops, the change was incorrect. per the spec, union is 83 and alias is 84.

File: zng/union.go (hunk 3 above)
Comment (henridf) [reply zng/union.go]: you bet. "pourquoi faire simple quand on peut faire compliqué" :)

File: zng/union.go (hunk 4 above)
Comment (henridf) [reply zng/union.go]: done

File: zng/union.go (hunk 5 above)
Comment (henridf) [reply zng/union.go]: ah, nice save.

**Comment by mattnibs** [comment]
I'm curious how union types will work in procs- groupby and sort especially. Will that be addressed in another pr?

**Comment by henridf** [comment]
@mattnibs there's more PRs to come and I don't have the answer to your question yet :)

**Pull request merged**